### PR TITLE
bridge: T3042: Support VLAN filter and VLAN sub-interface on the bridge

### PR DESCRIPTION
**Pull request by jack9603301**

### CODE REVIEW

**Comment by c-po** [comment]
I like to see someone working on this topic. Thanks You! - See my comments in Phabricator.

File: interface-definitions/interfaces-bridge.xml.in
Comment (c-po) [inline interface-definitions/interfaces-bridge.xml.in]: Why is this node required when we already have the vif nodes to include?

File: python/vyos/ifconfig/bridge.py
Comment (c-po) [inline python/vyos/ifconfig/bridge.py]: Adding whitespaces is not nice ;)

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/interfaces-bridge.xml.in
Comment (jack9603301) [reply interface-definitions/interfaces-bridge.xml.in]: The comment of the switch indicates that once the user sets it, the vlan filter processing function of the bridge will be enabled, and the bridge will become a vlan-aware bridge

Only when this switch is turned on, the settings of `member interface eth0 tagged` and `member interface eth0 untagged` can take effect

Comment (c-po) [reply interface-definitions/interfaces-bridge.xml.in]: Thy don't we make this auto recognized by the Python code once vif or vif-s or allowed-vlan or native-vlan option is set?

Comment (jack9603301) [reply interface-definitions/interfaces-bridge.xml.in]: Do you want to automatically enable vlan filter in this case instead of letting users specify?

Comment (c-po) [reply interface-definitions/interfaces-bridge.xml.in]: If a user configures all the required stuff for vlan-aware bridging we can safely assume he want's to have vlan aware bridging.

Comment (jack9603301) [reply interface-definitions/interfaces-bridge.xml.in]: I agree, and have modified the implementation

**Comment by c-po** [comment]
Please also do not forget to extend the current bridge smoketests with a new testcase validating this new feature.

https://github.com/vyos/vyos-1x/blob/current/smoketest/scripts/cli/test_interfaces_bridge.py

**Comment by jack9603301** [comment]
Basic smoke test has been submitted

File: smoketest/scripts/cli/test_interfaces_bridge.py
Comment (c-po) [inline smoketest/scripts/cli/test_interfaces_bridge.py]: Please add validation logic that the vlan interfaces have been created and also the allowed-vlan and native-vlan is properly setup. Right now this only tests the no error is generated by CLI but not if the functionality actually works.

File: python/vyos/validate.py
Comment (c-po) [inline python/vyos/validate.py]: This should be moved to `configdict.py` as this is a similar function as `is_member()`.

File: src/conf_mode/interfaces-bridge.py
Comment (c-po) [inline src/conf_mode/interfaces-bridge.py]: native_vlan and allowed_vlan share the same structure - please combine this in a single helper function which is called with the appropriate key (native_vlan or allowed_vlan)

**Comment by c-po** [comment]
Please incorporate the latest change requests - I run the smoketests and they passed locally - but the logic is actually not verified as only the CLI is tested for errors.

**Comment by jack9603301** [comment]
The verification logic I added for the time being only judges whether to activate the vlan filter processing mechanism. I think this is enough. If the processing mechanism is activated, it means that the bridge is a vlan-aware bridge.

Please read the comments of @runar in https://phabricator.vyos.net/T3042. Do you maintain the current command design?

**Comment by c-po** [comment]
I hesitate to merge this until there is a full validation logic, sorry. Testing becomes harder and harder for every new feature. Using automated unit or smoketests eases the life of every maintainer. It's write once test always thus it is very beneficial for the project.

**Comment by jack9603301** [comment]
> I hesitate to merge this until there is a full validation logic, sorry. Testing becomes harder and harder for every new feature. Using automated unit or smoketests eases the life of every maintainer. It's write once test always thus it is very beneficial for the project.

The main difficulty is that I don’t know how to parse `bridge -c vlan show`. This shows all the settings of vlan filter. The `vlan_filtering` flag will be automatically set after judging that `native_vlan` and `allowed_vlan` are set (according to the configuration file) , So I think this should be reliable, and the `vlan_filtering` flag indicates that the bridge is indeed running vlan filter, or can you give me a supplementary idea?

**Comment by jack9603301** [comment]
@c-po Just now, I found a way to parse the state through json. Now, I have updated the PR. However, because of the need to call `Poepn` to execute external commands during the test, there seems to be a warning from the SubProcess module

I have implemented the complete test logic on VLAN-aware bridges

File: interface-definitions/interfaces-bridge.xml.in
Comment (c-po) [inline interface-definitions/interfaces-bridge.xml.in]: Please shorten the help string, it makes my terminal explode ;).

```
3# set interfaces bridge br0 member interface eth1
Possible completions:
+  allowed-vlan Allow the bridge to pass the tagged VLAN tag on this interface
   cost         Bridge port cost
   native-vlan  Set the interface in ACCESS mode, automatically append vlan id to packets without the vlan id in the incoming traffic, and automatically strip the vlan id when exporting packets with specific vlan id
   priority     Bridge port priority
```

File: interface-definitions/interfaces-bridge.xml.in
Comment (c-po) [inline interface-definitions/interfaces-bridge.xml.in]: Why do I need to disable the native VLAN? When I do not define it it should be disabled. We should not add superfluous CLI nodes.

File: interface-definitions/interfaces-bridge.xml.in
Comment (c-po) [inline interface-definitions/interfaces-bridge.xml.in]: Please make this more user readable like the following:
```
Possible completions:
   <n>     Individual Virtual Local Area Network (VLAN) ID
   <n-m>   Virtual Local Area Network (VLAN) ID Range
```

Comment (c-po) [reply interface-definitions/interfaces-bridge.xml.in]: Please also note that the VLAN range is only 1-4094 as 0 and 4095 are reserved and can not be used. Also your above validator is restricted to this range.

File: src/op_mode/show_interfaces.py
Comment (c-po) [inline src/op_mode/show_interfaces.py]: Please do not change this - we should discuss the format in a dedicated task.

Comment (c-po) [reply src/op_mode/show_interfaces.py]: It actually breaks when no VLAN interfaces exist:

```
cpo@LR1.wue3# run show interfaces
Traceback (most recent call last):
  File "/usr/libexec/vyos/op_mode/show_interfaces.py", line 305, in <module>
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface        IP Address                        VLAN Mode S/L  Description
---------        ----------                        --------- ---  -----------
dum0             172.18.254.201/32                 Disabled  u/u
eth0             -                                 Disabled  u/u
eth0.201         172.18.201.10/24                  Disabled  u/u
    args.vrrp
  File "/usr/libexec/vyos/op_mode/show_interfaces.py", line 223, in run_show_intf_brief
    mode = [_ for _ in interface.get_vlan_mode_bind_bridge()]  or ['Disabled', ]
  File "/usr/lib/python3/dist-packages/vyos/ifconfig/interface.py", line 727, in get_vlan_mode_bind_bridge
    with open(f'/sys/class/net/{bridge_interface_name}/bridge/vlan_filtering', 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/sys/class/net/br0/bridge/vlan_filtering'
```

**Review by c-po — CHANGES_REQUESTED** [review]

File: smoketest/scripts/cli/test_interfaces_bridge.py
Comment (c-po) [inline smoketest/scripts/cli/test_interfaces_bridge.py]: Please re-use the `vyos.util.cmd()` wrapper for this and not re-implement reading from stdout.

File: smoketest/scripts/cli/test_interfaces_bridge.py
Comment (c-po) [inline smoketest/scripts/cli/test_interfaces_bridge.py]: Please assign the allowed-vlan range and native vlan to a variable which is referenced in the validation logic below.
Using this makes it easier to extend ranges or adding additional ranges via loops.

**Review by c-po — CHANGES_REQUESTED** [review]

File: src/op_mode/show_interfaces.py
Comment (jack9603301) [reply src/op_mode/show_interfaces.py]: I will cancel it in this implementation and archive the correction of the shoe interfaces locally for later use. But I don't understand why I can't find the VLAN filtering path of br0. Strange? This switch should be available for every bridge device and does pass the test on the local virtual machine

File: interface-definitions/interfaces-bridge.xml.in
Comment (jack9603301) [reply interface-definitions/interfaces-bridge.xml.in]: Because in the bridge with vlan_filter enabled by default, each port has a vlan 1 by default, `disable` will mean that a port will neither set `native-vlan` nor delete vlan 1

File: interface-definitions/interfaces-bridge.xml.in
Comment (jack9603301) [reply interface-definitions/interfaces-bridge.xml.in]: In fact, I originally used this validator, but because of some problems, the validation logic is now moved to the configuration script

**Comment by jack9603301** [comment]
> Why do I need to disable the native VLAN? When I do not define it it should be disabled. We should not add superfluous CLI nodes.

I have made some responses in the comments you provided to request changes, I need to wait for your further response to determine whether to change the details

File: interface-definitions/interfaces-bridge.xml.in
Comment (c-po) [reply interface-definitions/interfaces-bridge.xml.in]: This might be the underlaying implementation, I'm referring to the CLI. If there is no `native-vlan` node defined, it corresponds to `disable` thus you can get rid of the entire `disable` node.

Comment (jack9603301) [reply interface-definitions/interfaces-bridge.xml.in]: OK, I have updated the PR

**Comment by c-po** [comment]
@jack9603301 one last thing and then it looks good to me. In `test_interfaces_bridge.py` please enable the VLAN and QinQ testcases so they are executed from the base class by:

```
diff --git i/smoketest/scripts/cli/test_interfaces_bridge.py w/smoketest/scripts/cli/test_interfaces_bridge.py
index ed86d5fc..bba3e845 100755
--- i/smoketest/scripts/cli/test_interfaces_bridge.py
+++ w/smoketest/scripts/cli/test_interfaces_bridge.py
@@ -29,7 +29,8 @@ class BridgeInterfaceTest(BasicInterfaceTest.BaseTest):
         super().setUp()

         self._test_ipv6 = True
-
+        self._test_vlan = True
+        self._test_qinq = True
         self._base_path = ['interfaces', 'bridge']
         self._interfaces = ['br0']
```

**Review by c-po — APPROVED** [review]

**Comment by c-po** [comment]
@dmbaturin or @jestabro this looks good to me and it's a nice feature. Please provide a second pair of eyes on this CLI extension. Thanks!

**Comment by jestabro** [comment]
I notice one minor error in `helper_check_removed_vlan`: subscripting `key_mangling` when it is default None --- I realize that should not show up in practice, as this code is the only user, and key_mangling is rarely, if ever, None. Otherwise, it seems reasonable to me, and good work navigating some tricky parts of the code base.

**Comment by jack9603301** [comment]
@jestabro Thanks, I have revised the PR

**Comment by thomas-mangin** [comment]
nice PR, cosmetically, `tmp = dict_search('NAME')"` .. perhaps should be `name = dict_search('NAME')` and not `tmp` but it is a cosmetic comment only.

**Comment by jack9603301** [comment]
> nice PR, cosmetically, `tmp = dict_search('NAME')"` .. perhaps should be `name = dict_search('NAME')` and not `tmp` but it is a cosmetic comment only.

Need to be modified?

**Pull request merged**
